x-pack/plugin/vector-tile/src/main/java/org/elasticsearch/xpack/vectortile/rest/GridAggregation.java

```diff
@@ -0,0 +1,208 @@
+            if (z == 0 || gridPrecision == 0) {
+                // no need to buffer at level 0 as we are looking to all data.
+                return tile;
+            }
+            final int aggPrecision = gridPrecisionToAggPrecision(z, gridPrecision);
+            return new Rectangle(
+                GeoUtils.normalizeLon(tile.getMinX() - LON_BUFFER_SIZE[aggPrecision]),
+                GeoUtils.normalizeLon(tile.getMaxX() + LON_BUFFER_SIZE[aggPrecision]),
+                Math.min(GeoTileUtils.LATITUDE_MASK, tile.getMaxY() + LAT_BUFFER_SIZE[aggPrecision]),
+                Math.max(-GeoTileUtils.LATITUDE_MASK, tile.getMinY() - LAT_BUFFER_SIZE[aggPrecision])
+            );
+        }
+
+        @Override
+        public int gridPrecisionToAggPrecision(int z, int gridPrecision) {
+            // The minimum resolution we allow is 2 to avoid having hexagons containing the pole.
+            // this is the table of Hex precision:
+            // precision: 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 |
+            // -------------------------------------------
+            // zoom 0: 2 | 2 | 2 | 2 | 2 | 2 | 3 | 3 |
+            // zoom 1: 2 | 2 | 2 | 2 | 2 | 3 | 3 | 4 |
+            // zoom 2: 2 | 2 | 2 | 2 | 3 | 3 | 4 | 4 |
+            // zoom 3: 2 | 2 | 2 | 3 | 3 | 4 | 4 | 5 |
+            // zoom 4: 2 | 2 | 3 | 3 | 4 | 4 | 5 | 5 |
+            // zoom 5: 2 | 3 | 3 | 4 | 4 | 5 | 5 | 6 |
+            // zoom 6: 3 | 3 | 4 | 4 | 5 | 5 | 6 | 6 |
+            // zoom 7: 3 | 4 | 4 | 5 | 5 | 6 | 6 | 7 |
+            // zoom 8: 4 | 4 | 5 | 5 | 6 | 6 | 7 | 7 |
+            // zoom 9: 4 | 5 | 5 | 6 | 6 | 7 | 7 | 8 |
+            // zoom 10: 5 | 5 | 6 | 6 | 7 | 7 | 8 | 8 |
+            // zoom 11: 5 | 6 | 6 | 7 | 7 | 8 | 8 | 9 |
+            // zoom 12: 6 | 6 | 7 | 7 | 8 | 8 | 9 | 9 |
+            // zoom 13: 6 | 7 | 7 | 8 | 8 | 9 | 9 | 10 |
+            // zoom 14: 7 | 7 | 8 | 8 | 9 | 9 | 10 | 10 |
+            // zoom 15: 7 | 8 | 8 | 9 | 9 | 10 | 10 | 11 |
+            // zoom 16: 8 | 8 | 9 | 9 | 10 | 10 | 11 | 11 |
+            // zoom 17: 8 | 9 | 9 | 10 | 10 | 11 | 11 | 12 |
+            // zoom 18: 9 | 9 | 10 | 10 | 11 | 11 | 12 | 12 |
+            // zoom 19: 9 | 10 | 10 | 11 | 11 | 12 | 12 | 13 |
+            // zoom 20: 10 | 10 | 11 | 11 | 12 | 12 | 13 | 13 |
+            // zoom 21: 10 | 11 | 11 | 12 | 12 | 13 | 13 | 14 |
+            // zoom 22: 11 | 11 | 12 | 12 | 13 | 13 | 14 | 14 |
+            // zoom 23: 11 | 12 | 12 | 13 | 13 | 14 | 14 | 15 |
+            // zoom 24: 12 | 12 | 13 | 13 | 14 | 14 | 15 | 15 |
+            // zoom 25: 12 | 13 | 13 | 14 | 14 | 15 | 15 | 15 |
+            // zoom 26: 13 | 13 | 14 | 14 | 15 | 15 | 15 | 15 |
+            // zoom 27: 13 | 14 | 14 | 15 | 15 | 15 | 15 | 15 |
+            // zoom 28: 14 | 14 | 15 | 15 | 15 | 15 | 15 | 15 |
+            // zoom 29: 14 | 15 | 15 | 15 | 15 | 15 | 15 | 15 |
+            return Math.min(H3.MAX_H3_RES, Math.max(2, (z + gridPrecision - 1) / 2));
```

Comment:
You are right that the hex bins resolution for high zooms is too low and can be improved. Still what you propose would not wok either because the density of hex bin in your example is too big.

For geotile and grid_precision 8 we can generate up to 65536 bins which is the default limit for the number of buckets in a bucket aggregation. Therefore the density of hex bins should always be <= density of geotile, otherwise we might go over the limit of the number of buckets.

In your proposal, the density of hex bins can be up to 2.5 times higher than tile bins.  We should make sure density is not bigger than 1 (it can be slightly bigger as the number of hex bins consider hexes around the poles which are not part of the mercator projection and for small zoom level when the number of bins < 65536).

Here is my approximation, what I wonder is how to document this.

<img width="728" alt="image" src="https://user-images.githubusercontent.com/29038686/157433901-6d37592b-7b22-46fa-8029-41c92d49a3d9.png">

Reply:
I updated how we compute the H3 resolution by using the table above. Let me know what you think @thomasneirynck 
@craigtaverner